### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by NaleRaphael**
Hi @davidtvs, this is a draft of unit tests for this package. Let's continue the discussion and get it completed! And please feel free to give me any suggestion.

Currently, I build some classes inheriting `BaseTask` to make us able to write test cases more conveniently. And I choose `pytest` as the unit test framework, since there are bunch of functional extensions which may help us maintain these test easily. You can check out more details in `tests/README.md`.

There are still some tests need to be done, I will finish them in these days.

Besides, here are some items that I would recommend to do later:
- [ ] CI/CD (`travis-ci` is a good choice for GitHub project)
- [ ] code coverage report (use `pytest-cov` and host the report on `codecov.io`)

Finally, [a compatibility issue](https://github.com/NaleRaphael/pytorch-lr-finder/commit/9e8796a) is found during the development of these tests. But I'm not sure whether it is good to send it as a new PR. In order to make it be managed easily, it's attached in this PR currently.

### CODE REVIEW

File: tests/task.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.utils.data import DataLoader, Subset
+
+from .model import LinearMLP
+from .dataset import XORDataset, ExtraXORDataset
+
+
+def use_cuda():
+    if int(os_getenv('CPU_ONLY', '0')):
```
Comment (davidtvs) [inline tests/task.py:13]: This is a good idea. A possible improvement would be to make it a pytest command-line argument to avoid using environment variables

File: tests/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    ```bash
+    # in root directory of this package
+    $ python -mpytest ./tests
+    ```
+
+- run tests on a machine without GPU
+    ```bash
+    # in root directory of this package
+    $ CPU_ONLY=1 python -mpytest ./tests
+    ```
```
Comment (davidtvs) [inline tests/README.md:15]: This is a good idea. A possible improvement would be to make it a pytest command-line argument instead of using environment variables.

Also, I think 'run tests on a machine without GPU' doesn't reflect the use-case. On a machine without GPU the normal execution will automatically pick the CPU. The use-case is more a  'force CPU-only execution'

**Comment by davidtvs** [comment]
About Python 2: since in the meantime Python 2.7 is reached end of life, I think we should just remove support for it. At this point, it's just a source of pain. I'll remove everything related to it today or the coming days.

About next steps: ya, I agree that both of those should be next. I'll probably add some github actions to check the code style too. Wanted to play around with that feature for some time now and this is a good excuse 😄 

I think this is a good initial set of tests for a PR, we can add more later and keep the PRs small and low complexity.

File: requirements-dev.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+pytest
```
Comment (davidtvs) [inline requirements-dev.txt:1]: We can just place this into the requirements.txt.

The dependencies listed in `setuptools.py` are for the end-user and the ones in the requirements.txt for local development.

**Comment by davidtvs** [comment]
Just made a PR to remove Python 2 support (#28). @NaleRaphael can you give it a look just in case I missed something.

We would then first merge #28 and then rebase this PR and remove the fix to the compatibility issue

File: tests/task.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        if not use_cuda():
+            self.device = None
+        if self.device is None:
+            return
+
+        if isinstance(self.device, str):
+            try:
+                self.device = torch.device(self.device)
+            except RuntimeError as ex:
+                raise ex
```
[the remark this answers is not in the document]
Comment (NaleRaphael) [reply tests/task.py:59]: Oh! I was trying to validate the device string by calling `torch.device()` and re-raise a `ValueError` when given device string is not supported. Like this:
```python
try:
    self.device = torch.device(self.device)
except RuntimeError as ex:
    raise ValueError('Invalid device string.') from ex
```
By doing this, we don't need to maintain a list for supported devices here.

But since I found that syntax of `raise ex2 from ex1` isn't supported in Py2k, I rewrote that line.
However, it seems the try-except block is redundant here. I'll remove the try-except block in the following commits.

File: tests/README.md (hunk 2 above)
Comment (NaleRaphael) [reply tests/README.md:15]: Ya, you're right. `CPU_ONLY` is more likely a term indicating 'force CPU-only execution'. I'll revise the description.

And it seems nice to replace it with a pytest command-line argument. But since it isn't used by fixtures directly, I'm planning to add a custom config object to store any further command line arguments, and bind it to `pytest` namespace dynamically. I found this approach works since [`pytest.config` global object is deprecated in pytest 5.0](https://docs.pytest.org/en/latest/deprecations.html#pytest-config-global).

```python
# In conftest.py

def pytest_addoption(parser):
    parser.addoption(
        '--cpu_only', action='store_true',
        help='Forcibly run all tests on CPU.'
    )

def pytest_configure(config):
    # bind a config object to `pytest`
    pytest.custom_cmdopt = CustomCommandLineOption()

    # add command line argument to that config object
    pytest.custom_cmdopt.add(
        'cpu_only', config.getoption('--cpu_only')
    )
```

```python
# In task.py
import pytest

def use_cuda():
    if pytest.custom_cmdopt.cpu_only:
        return False
    else:
        return torch.cuda.is_available()
```

File: requirements-dev.txt (hunk 3 above)
Comment (NaleRaphael) [reply requirements-dev.txt:1]: OK, I'll remove this file!

File: tests/task.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NaleRaphael) [reply tests/task.py:59]: Since we are going to remove support of Py2k, I think it is still a good idea to remove this try-except block. Let's just raise that `RuntimeError` in a fail-fast style, and it would be a clear error message for developers like us.

**Comment by NaleRaphael** [comment]
The fix for Py2k compatibility issue is removed. I'll submit a revision for requested changes later 😎

File: tests/task.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            self.train_loader = DataLoader(dataset)
+            self.val_loader = None
+
+        self.model = LinearMLP([8, 4, 1])
+        self.optimizer = optim.SGD(self.model.parameters(), lr=1e-3)
+        self.criterion = nn.MSELoss()
+        self.device = torch.device('cuda')
+
+
+class DifferentLearningRateTask(BaseTask):
```
Comment (davidtvs) [inline tests/task.py:101]: I suggest using `Differential` or `Discriminative` instead of `Different`. Differential is the popular name for this but discriminative is the `official` name; so either one is fine

Comment (NaleRaphael) [reply tests/task.py:101]: Good point, I was a bit confused while trying to name this task.
It seems `Discriminative` is a good choice for me after seeing [this comment](https://forums.fast.ai/t/what-happened-to-differential-learning-rates-and-warmup/28809/3).

File: tests/test_lr_finder.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import pytest
+from torch_lr_finder import LRFinder
+
+from . import task as mod_task
+
+
+def collect_task_classes():
+    from . import task as mod
+    from .task import BaseTask
```
Comment (davidtvs) [inline tests/test_lr_finder.py:9]: Any reason for the local imports? I think these can be deleted and in the function replace `task` with `mod_task` and `BaseTask` with `mod_task.BaseTask` (except for the string 'BaseTask' of course)

Comment (NaleRaphael) [reply tests/test_lr_finder.py:9]: You are right. These lines were for keeping global namespace clean while I intended to make all test cases able to be re-run automatically with various tasks. But since I found it not necessary to do that for some cases (e.g. `test_reset`) and add `from . import task as mod_task` at the top of script, it's fine to remove these lines.
Thanks for pointing this out. :)

**Comment by NaleRaphael** [comment]
BTW, would it be better to format these code with a linter you are using? I could apply the configuration you given and submit a new commit for fixing the coding style, and it might reduce your workload on checking it. 😊
And once you think this PR is ready, please feel free to squeeze these commits to make it smaller. (Or I could do that from my side)

**Comment by davidtvs** [comment]
Just pushed my flake8 configuration. For formatting, I use [black](https://github.com/psf/black) with default settings.

**Comment by NaleRaphael** [comment]
I've updated this branch and formatted the code with black. And the following part is kept as it was in case there are more tasks to be run in the future:
```python
# in test_lr_finder.py
@pytest.mark.parametrize(
    "cls_task",
    [
        mod_task.XORTask,
        mod_task.DiscriminativeLearningRateTask,
    ],
)
def test_reset(self, cls_task):
```

**Comment by davidtvs** [comment]
Looks good, thanks for the PR.

**Comment by NaleRaphael** [comment]
Thanks! More tests would be done lately in a few weeks to make the coverage rate higher, and I would open a new PR when it is ready.
